intro-to-R-tidyverse/03-intro_to_tidyverse.Rmd

```diff
@@ -387,36 +387,102 @@ In base R, the `apply` family of functions can be an alternative methods for
 performing transformations across a data frame, matrix or other object structures. 
 
 One of this family is (shockingly) the function `apply()`, which operates on 
-matrices or data frames. 
+matrices.
+
+A matrix is similar to a data frame in that it is a rectangular table of data,
+but it has an additional constraint: 
+rather than each column having a type, ALL data in a matrix has the same type.
+
 The first argument to `apply()` is the data object we want to work on.
 The third argument is the function we will apply to each row or column of the 
 data object.
 The second argument in specifies whether we are applying the function 
 across rows or across columns (1 for rows, 2 for columns).
 
-Remember that `gene_df` is a gene x sample gene expression data frame.
+Remember that `gene_df` is a gene x sample gene expression data frame that has
+columns of two different types, character and numeric.
+Converting it to a matrix will require us to make them all the same type.
+We can coerce it into a matrix using `as.matrix()`, in which case R will
+pick a type that it can convert everything to.
+What does it choose?
+
+```{r matrix}
+# Coerce `gene_df` into a matrix
+gene_matrix <- as.matrix(gene_df)
+```
+
+```{r matrix-type, live = TRUE}
+# Explore the structure of the `gene_matrix` object
+str(gene_matrix)
+```
+
+While that worked, it is rare that we want numbers converted to text, so we are
+going to select only the columns with numeric values before converting it to a matrix.
+We can do this most easily by removing the first column, which contains the gene names 
+stored as character values.
+
+```{r matrix-numeric, live = TRUE}
+# Let's save a new matrix object names `gene_num_matrix` containing only
+# the numeric values
+gene_num_matrix <- as.matrix(gene_df[, -1])
+
+# Explore the structure of the `gene_num_matrix` object
+str(gene_num_matrix)
+```
+
+Why do we have a `[, -1]` after `gene_df` in the above chunk?
+
+Now that the matrix is all numbers, we can do things like calculate the column
+or row statistics using `apply()`.
 
 ```{r rowmeans}
 # Calculate row means
-gene_means <- apply(gene_df[, -1], 1, mean) # Notice we are using 1 here
+gene_means <- apply(gene_num_matrix, 1, mean) # Notice we are using 1 here
 
 # How long will `gene_means` be? 
 length(gene_means)
 ```
 
-Why do we have a `[, -1]` after `gene_df` in the above chunk?
+Note that we can obtain the same results using the columns with numeric values
+from the `gene_df` data frame.
+
+```{r rowmeans-dataframe}
+# Calculate row means using the `gene_df` object
+gene_means_from_df <- apply(gene_df[, -1], 1, mean) # Removing the character
+# column allows R to do the as.matrix() coercion automatically, which only works
+# if the rest of the data frame is numeric.
```

Comment:
```suggestion
Note that we can obtain the same results if we select just the columns with numeric
values from the `gene_df` data frame.
This allows R to do the as.matrix() coercion automatically, and can be a handy shortcut
if you have a *mostly* numeric data frame. 

```{r rowmeans-dataframe}
# Calculate row means using the `gene_df` object after removing the character column
# apply() converts this to a matrix internally
gene_means_from_df <- apply(gene_df[, -1], 1, mean) 
```